### DISCUSSION TITLE
How to enable printf/log in contracts written in C

### DISCUSSION THREAD

**Discussion by XuJiandong** [nervosnetwork/ckb-vm#193] [General]
1. set  [-fno-builtin-printf](https://github.com/nervosnetwork/ckb-production-scripts/blob/ed06f45d54b491e4621a339e3d7e1f9b3a771e24/Makefile#L5) while compiling C code
2. Add macro [CKB_C_STDLIB_PRINTF](https://github.com/nervosnetwork/ckb-production-scripts/blob/ed06f45d54b491e4621a339e3d7e1f9b3a771e24/c/xudt_rce.c#L2-L4) before including stdio.h.
3. With Rust test cases, [should specify how to receive the log](https://github.com/nervosnetwork/ckb-production-scripts/blob/ed06f45d54b491e4621a339e3d7e1f9b3a771e24/tests/xudt_rce_rust/tests/test_xudt_rce.rs#L1302)
```Rust
fn debug_printer(script: &Byte32, msg: &str) {
    let slice = script.as_slice();
    let str = format!(
        "Script({:x}{:x}{:x}{:x}{:x})",
        slice[0], slice[1], slice[2], slice[3], slice[4]
    );
    println!("{:?}: {}", str, msg);
}
verifier.set_debug_printer(debug_printer);
```

**Comment by xxuejie** [comment]
Note that in CKB contract development, we consider it a best practice to always add `-fno-builtin-printf` when building C smart contracts.
